### ISSUE TITLE
Infinite loop on driver open fail

### ISSUE THREAD

**Issue by marcucio** [mozilla/localForage#511]
fix for bug where we were going into an infinite loop if the open database failed (saw this on BB10 cordova app)

I don't think we should call _initStorage again since it just failed

fixed in pull request: #510

**Comment by thgreasi** [comment]
Perhaps we could remove this fallback and just reject the promise. The mechanism in the main localforage.js module can handle the fallback to the next preferred driver. I had seen that earlier, before the es6 refactor and started a discussion about it in an other issue, but can't find it right now. This was probably used before the fallback-on-fail mechanism was introduced, to prevent some missconfigurations.

**Comment by marcucio** [comment]
@thgreasi so your saying we might not have to do self.setDriver(self.LOCALSTORAGE), it looks like indexdb does it this way (just resolves it):

```
        function ignoreErrors() {
            // Don't handle errors here,
            // just makes sure related localForages aren't pending.
            return Promise.resolve();
        }

        for (var j = 0; j < dbContext.forages.length; j++) {
            var forage = dbContext.forages[j];
            if (forage !== self) { // Don't wait for itself...
                initPromises.push(forage._initReady().catch(ignoreErrors));
            }
        }

```

**Comment by thgreasi** [comment]
Exactly. I think that catch should just do a reject().
Any chance you can provide a test case or a code example with the infinite
loop?

> --
*Thodoris Greasidis*
Computer, Networking &
Communications Engineer

**Comment by marcucio** [comment]
yes, it is easy to repro:
in websql.js replace:
```
dbInfo.db = openDatabase(dbInfo.name, String(dbInfo.version), dbInfo.description, dbInfo.size);
``
with:

```
throw new Error('quota exceded');
```

results in this:
```
Running "mocha:unit" (mocha) task
Testing: http://localhost:9999/test/test.component.html

  
  ․
Running PhantomJS...ERROR
>> PhantomJS has crashed. Please read the crash reporting guide at https://github.com/ariya/phantomjs/wiki/Crash-Reporting and file a bug report at https://github.com/ariya/phantomjs/issues/new with the crash dump file attached: /tmp/4966558E-41D1-407A-8444-C4363882E5FC.dmp 0 [ 'PhantomJS has crashed. Please read the crash reporting guide at https://github.com/ariya/phantomjs/wiki/Crash-Reporting and file a bug report at https://github.com/ariya/phantomjs/issues/new with the crash dump file attached: /tmp/4966558E-41D1-407A-8444-C4363882E5FC.dmp' ]
Warning: PhantomJS exited unexpectedly with exit code 2. Use --force to continue.

Aborted due to warnings.
localForage mike$ 
```

**Comment by thgreasi** [comment]
I guess we will have to add a test with since monkey patching...

On Tue, Jan 26, 2016, 21:13 Mike <notifications@github.com> wrote:

> yes, it is easy to repro:
> in websql.js replace:
>
> dbInfo.db = openDatabase(dbInfo.name, String(dbInfo.version), dbInfo.description, dbInfo.size);
> ``
> with:
>
>
> throw new Error('quota exceded');
>
>
> results in this:
>
> Running "mocha:unit" (mocha) task
> Testing: http://localhost:9999/test/test.component.html
>
> ․
> Running PhantomJS...ERROR
>
> PhantomJS has crashed. Please read the crash reporting guide at
> https://github.com/ariya/phantomjs/wiki/Crash-Reporting and file a bug
> report at https://github.com/ariya/phantomjs/issues/new with the crash
> dump file attached: /tmp/4966558E-41D1-407A-8444-C4363882E5FC.dmp 0 [
> 'PhantomJS has crashed. Please read the crash reporting guide at
> https://github.com/ariya/phantomjs/wiki/Crash-Reporting and file a bug
> report at https://github.com/ariya/phantomjs/issues/new with the crash
> dump file attached: /tmp/4966558E-41D1-407A-8444-C4363882E5FC.dmp' ]
> Warning: PhantomJS exited unexpectedly with exit code 2. Use --force to
> continue.
>
> Aborted due to warnings.
> localForage mike$
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/localForage/issues/511#issuecomment-175185756>
> .
>
-- 
*Thodoris Greasidis*
Computer, Networking &
Communications Engineer

**Comment by marcucio** [comment]
We might be able to redefine window.openDatabase to ‘undefined’ so that it throws an exception during a test, not sure I will have to play around with it. I’ll be pretty busy today, maybe tomorrow I can look into it

--
Mike Marcucio
marcucio.com <http://marcucio.com/>
mike@marcucio.com <mailto:mike@marcucio.com>

blog <http://blog.getitdoneapp.com/> | twitter <https://twitter.com/GetItDoneBlog> | facebook <https://www.facebook.com/GetItDoneTasks> | google+ <https://plus.google.com/u/0/b/116698933997097774721/+Marcucio/posts>

**Comment by thgreasi** [comment]
Or even better to a function that throws an error.
Something similar is done in [test.nodriver.html](https://github.com/mozilla/localForage/blob/master/test/test.nodriver.html).
How about placing the new tests in a new file named `test.faultydrivers.html` or `test.drivererrors.html`?

**Comment by tofumatt** [comment]
Merged in #517 with some tidying.
